# 04 · Reversible, pushed to the maximum batch

First the ceiling, for both stacks: each candidate batch trains 4 steps in a fresh process, with the MPS
allocator capped at the recommended working set (so out-of-memory is a clean failure, not swap). The
search doubles until a batch fails and then bisects.

Then the reversible model trains the same 50M tokens at its own ceiling. A bigger batch means fewer
optimizer steps for the same tokens, so the peak learning rate is scaled by sqrt(batch / 32) (the usual
square-root rule for Adam), with 5% warmup. Everything else is unchanged.

In [1]:
import sys, os, json, pathlib, subprocess
IN_COLAB = "google.colab" in sys.modules
if IN_COLAB and not pathlib.Path("ERA_V5").exists():
    subprocess.run(["git", "clone", "--depth", "1", "https://github.com/amukka/ERA_V5.git"], check=True)
    os.chdir("ERA_V5/session13_Distributed_Training")
ROOT = pathlib.Path.cwd()
while not (ROOT / "src").exists() and ROOT != ROOT.parent:
    ROOT = ROOT.parent
os.chdir(ROOT); sys.path.insert(0, str(ROOT))
if not (ROOT / "data" / "train.bin").exists():          # ~5 minutes: stream FineWeb-Edu, train the BPE
    subprocess.run([sys.executable, "tools/prepare_data.py"], check=True)

import torch, platform
from src import experiments as X
from src.train import pick_device
dev = pick_device()
print("device:", dev, "| torch", torch.__version__, "| python", platform.python_version(), "|", platform.platform())
if dev == "mps":
    print(f"MPS recommended working set: {torch.mps.recommended_max_memory()/2**30:.2f} GiB")
if dev == "cuda":
    print(torch.cuda.get_device_name(), f"{torch.cuda.get_device_properties(0).total_memory/2**30:.1f} GiB")
print(json.dumps(json.loads((ROOT / "data" / "meta.json").read_text()), indent=1))

device: mps | torch 2.13.0 | python 3.14.5 | macOS-26.6.2-arm64-arm-64bit-Mach-O
MPS recommended working set: 11.84 GiB
{
 "source": "HuggingFaceFW/fineweb-edu, config sample-10BT, streamed in order",
 "vocab_size": 8192,
 "eos_id": 0,
 "train_tokens": 56000000,
 "train_docs": 46080,
 "val_tokens": 1000000,
 "val_docs": 1024,
 "chars_per_token": 3.886,
 "note": "val.bin comes from the tokenizer's training documents; train.bin from the documents after them",
 "seconds": 314.3
}


**The model** (identical in every run): GPT, 10 layers, d_model 384, 6 heads, MLP 4x, sequence 512,
byte-level BPE vocabulary 8,192 with the head tied to the embedding. **21,053,184 parameters**, of which
17,710,848 are outside the embedding tables. No dropout anywhere, because a reversible backward pass
recomputes each block and a random mask would recompute differently. AdamW (0.9, 0.95, wd 0.1), peak LR
1e-3, 3% warmup then cosine to 10%, gradient clip 1.0, fp32.

In [2]:
chosen = json.loads((X.RESULTS / "chosen_variant.json").read_text())
print("searching the standard stack")
max_std, trail_std = X.max_batch("standard", 1.0, start=32, step=4)
print("searching the reversible stack:", chosen["variant"], "h =", chosen["h"])
max_rev, trail_rev = X.max_batch(chosen["variant"], chosen["h"], start=32, step=8)
print(f"\nmax batch  standard: {max_std}   {chosen['variant']}: {max_rev}   ratio {max_rev/max_std:.2f}x")
(X.RESULTS / "max_batch.json").write_text(json.dumps(
    {"standard": {"max_batch": max_std, "trail": trail_std},
     chosen["variant"]: {"max_batch": max_rev, "trail": trail_rev, "h": chosen["h"]}}, indent=2))

searching the standard stack


  standard  batch   32  peak    9,466 MiB   10,550 tok/s


  standard  batch   64  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 12.02 GiB, other a


  standard  batch   48  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 11.77 GiB, other a


  standard  batch   40  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 10.90 GiB, other a


  standard  batch   36  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 10.86 GiB, other a


searching the reversible stack: momentum h = 0.5


  momentum  batch   32  peak    2,754 MiB    8,972 tok/s


  momentum  batch   64  peak    5,986 MiB    9,023 tok/s


  momentum  batch  128  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 9.03 GiB, other al


  momentum  batch   96  does not fit: RuntimeError: MPS backend out of memory (MPS allocated: 10.03 GiB, other a


  momentum  batch   80  peak    6,754 MiB    7,860 tok/s


  momentum  batch   88  peak    8,157 MiB    7,745 tok/s



max batch  standard: 32   momentum: 88   ratio 2.75x


3332

In [3]:
import math
lr = 1e-3 * math.sqrt(max_rev / X.BASE_BATCH)
print(f"batch {max_rev} = {max_rev*512:,} tokens a step, {X.TOKENS // (max_rev*512):,} steps, peak LR {lr:.2e}")
big = X.train(name=f"reversible_{chosen['variant']}_maxbatch", variant=chosen["variant"], h=chosen["h"],
              batch=max_rev, tokens=X.TOKENS, lr=lr, warmup_frac=0.05, mem_fraction=1.0,
              eval_every=max(1, 250 * X.BASE_BATCH // max_rev), log_every=2)

batch 88 = 45,056 tokens a step, 1,109 steps, peak LR 1.66e-03


{"name": "reversible_momentum_maxbatch", "device": "mps", "config": {"vocab_size": 8192, "seq_len": 512, "n_layer": 10, "n_head": 6, "d_model": 384, "variant": "momentum", "h": 0.5, "fp_iters": 6, "extra": {}}, "params": 21053184, "params_non_embedding": 17710848, "batch": 88, "tokens_per_step": 45056, "steps": 1109, "peak_lr": 0.0016583123951777, "warmup": 55, "amp_bf16": 0, "torch": "2.13.0", "machine": "arm64", "processor": "arm"}


step     0/1109  tok   0.05M  loss 9.0855  gnorm 3.374  0 tok/s  0.1 min


memory audit at step 3: exact peak 6,354 MiB in aten._log_softmax_backward_data.default


step    20/1109  tok   0.90M  loss 7.3019  gnorm 0.592  7,692 tok/s  2.0 min


step    40/1109  tok   1.80M  loss 6.7522  gnorm 0.623  7,727 tok/s  3.9 min


step    60/1109  tok   2.70M  loss 6.3700  gnorm 0.396  7,787 tok/s  5.8 min


step    80/1109  tok   3.60M  loss 6.1472  gnorm 0.329  7,737 tok/s  7.8 min


            eval @ 4.01M tokens: val loss 6.0855


step   100/1109  tok   4.51M  loss 5.9982  gnorm 0.298  7,680 tok/s  9.7 min


step   120/1109  tok   5.41M  loss 5.9237  gnorm 0.310  7,808 tok/s  11.6 min


step   140/1109  tok   6.31M  loss 5.8260  gnorm 0.296  7,838 tok/s  13.6 min


step   160/1109  tok   7.21M  loss 5.6651  gnorm 0.346  7,820 tok/s  15.5 min


            eval @ 8.07M tokens: val loss 5.6102


step   180/1109  tok   8.11M  loss 5.5962  gnorm 0.300  8,364 tok/s  17.4 min


step   200/1109  tok   9.01M  loss 5.5325  gnorm 0.350  7,813 tok/s  19.3 min


step   220/1109  tok   9.91M  loss 5.4298  gnorm 0.325  7,791 tok/s  21.2 min


step   240/1109  tok  10.81M  loss 5.3841  gnorm 0.368  7,760 tok/s  23.2 min


step   260/1109  tok  11.71M  loss 5.2825  gnorm 0.325  7,688 tok/s  25.1 min


            eval @ 12.12M tokens: val loss 5.3055


step   280/1109  tok  12.62M  loss 5.2584  gnorm 0.325  7,828 tok/s  27.0 min


step   300/1109  tok  13.52M  loss 5.2047  gnorm 0.346  7,788 tok/s  29.0 min


step   320/1109  tok  14.42M  loss 5.1208  gnorm 0.372  7,820 tok/s  30.9 min


step   340/1109  tok  15.32M  loss 5.1315  gnorm 0.359  7,783 tok/s  32.8 min


            eval @ 16.18M tokens: val loss 5.0506


step   360/1109  tok  16.22M  loss 5.0285  gnorm 0.364  8,361 tok/s  34.7 min


step   380/1109  tok  17.12M  loss 4.9724  gnorm 0.373  7,801 tok/s  36.7 min


step   400/1109  tok  18.02M  loss 4.9128  gnorm 0.368  7,818 tok/s  38.6 min


step   420/1109  tok  18.92M  loss 4.9409  gnorm 0.406  7,759 tok/s  40.5 min


step   440/1109  tok  19.82M  loss 4.8483  gnorm 0.408  7,713 tok/s  42.5 min


            eval @ 20.23M tokens: val loss 4.8514


step   460/1109  tok  20.73M  loss 4.8092  gnorm 0.359  7,703 tok/s  44.4 min


step   480/1109  tok  21.63M  loss 4.7430  gnorm 0.392  7,753 tok/s  46.3 min


step   500/1109  tok  22.53M  loss 4.7460  gnorm 0.387  7,757 tok/s  48.3 min


step   520/1109  tok  23.43M  loss 4.7177  gnorm 0.383  7,806 tok/s  50.2 min


            eval @ 24.29M tokens: val loss 4.6994


step   540/1109  tok  24.33M  loss 4.6724  gnorm 0.410  8,372 tok/s  52.1 min


step   560/1109  tok  25.23M  loss 4.6388  gnorm 0.410  7,775 tok/s  54.0 min


step   580/1109  tok  26.13M  loss 4.6218  gnorm 0.389  7,772 tok/s  56.0 min


step   600/1109  tok  27.03M  loss 4.5600  gnorm 0.403  7,789 tok/s  57.9 min


step   620/1109  tok  27.93M  loss 4.5900  gnorm 0.461  7,764 tok/s  59.9 min


            eval @ 28.34M tokens: val loss 4.5623


step   640/1109  tok  28.84M  loss 4.5481  gnorm 0.455  7,820 tok/s  61.8 min


step   660/1109  tok  29.74M  loss 4.4532  gnorm 0.474  7,825 tok/s  63.8 min


step   680/1109  tok  30.64M  loss 4.4765  gnorm 0.433  7,824 tok/s  65.7 min


step   700/1109  tok  31.54M  loss 4.4309  gnorm 0.387  7,770 tok/s  67.6 min


            eval @ 32.40M tokens: val loss 4.4574


step   720/1109  tok  32.44M  loss 4.4120  gnorm 0.437  8,355 tok/s  69.6 min


step   740/1109  tok  33.34M  loss 4.4041  gnorm 0.461  7,696 tok/s  71.5 min


step   760/1109  tok  34.24M  loss 4.4307  gnorm 0.497  7,791 tok/s  73.4 min


step   780/1109  tok  35.14M  loss 4.3695  gnorm 0.442  7,774 tok/s  75.4 min


step   800/1109  tok  36.04M  loss 4.3184  gnorm 0.484  7,710 tok/s  77.3 min


            eval @ 36.45M tokens: val loss 4.3624


step   820/1109  tok  36.95M  loss 4.3528  gnorm 0.454  7,751 tok/s  79.2 min


step   840/1109  tok  37.85M  loss 4.3145  gnorm 0.480  7,826 tok/s  81.1 min


step   860/1109  tok  38.75M  loss 4.3226  gnorm 0.485  7,786 tok/s  83.1 min


step   880/1109  tok  39.65M  loss 4.3296  gnorm 0.464  7,736 tok/s  85.0 min


            eval @ 40.51M tokens: val loss 4.2854


step   900/1109  tok  40.55M  loss 4.2608  gnorm 0.488  8,404 tok/s  86.9 min


step   920/1109  tok  41.45M  loss 4.2336  gnorm 0.524  7,808 tok/s  88.8 min


step   940/1109  tok  42.35M  loss 4.2096  gnorm 0.513  7,819 tok/s  90.8 min


step   960/1109  tok  43.25M  loss 4.1972  gnorm 0.507  7,732 tok/s  92.7 min


step   980/1109  tok  44.15M  loss 4.1926  gnorm 0.480  7,823 tok/s  94.6 min


            eval @ 44.56M tokens: val loss 4.2302


step  1000/1109  tok  45.06M  loss 4.2078  gnorm 0.539  7,810 tok/s  96.6 min


step  1020/1109  tok  45.96M  loss 4.2085  gnorm 0.470  7,682 tok/s  98.5 min


step  1040/1109  tok  46.86M  loss 4.1628  gnorm 0.488  7,761 tok/s  100.4 min


step  1060/1109  tok  47.76M  loss 4.1951  gnorm 0.463  7,797 tok/s  102.4 min


            eval @ 48.62M tokens: val loss 4.1898


step  1080/1109  tok  48.66M  loss 4.0926  gnorm 0.535  8,369 tok/s  104.3 min


step  1100/1109  tok  49.56M  loss 4.0861  gnorm 0.487  7,733 tok/s  106.2 min


step  1108/1109  tok  49.92M  loss 4.0923  gnorm 0.495  7,800 tok/s  107.0 min


            eval @ 49.92M tokens: val loss 4.1808


reversibility at final weights: grad rel err 8.69e-05, cosine 1.000000, state rebuild err 6.50e-05
{"tok_per_s": 7776.4, "final_train_loss": 4.092309951782227, "final_val_loss": 4.180837237834931, "peak_exact_mib": 6354.4, "peak_alloc_mib": 3539.4, "peak_driver_mib": 11555.3, "train_seconds": 6419.7}


In [4]:
keys = ["batch", "tokens_trained", "steps", "peak_lr", "final_train_loss", "final_train_loss_avg_last5pct",
        "final_val_loss", "tok_per_s", "train_seconds", "peak_exact_mib", "peak_exact_op", "peak_alloc_mib", "peak_driver_mib"]
print(json.dumps({k: big[k] for k in keys}, indent=1))
print(json.dumps(big["reversibility"], indent=1))

{
 "batch": 88,
 "tokens_trained": 49922048,
 "steps": 1109,
 "peak_lr": 0.0016583123951777,
 "final_train_loss": 4.092309951782227,
 "final_train_loss_avg_last5pct": 4.162381719659876,
 "final_val_loss": 4.180837237834931,
 "tok_per_s": 7776.4,
 "train_seconds": 6419.7,
 "peak_exact_mib": 6354.4,
 "peak_exact_op": "aten._log_softmax_backward_data.default",
 "peak_alloc_mib": 3539.4,
 "peak_driver_mib": 11555.3
}
{
 "init": {
  "grad_rel_err": 2.200799507637422e-06,
  "grad_cosine": 0.9999999999972711,
  "state_rebuild_rel_err_per_layer": [
   5.324042831489351e-06,
   2.7300793590256944e-06,
   1.537154957986786e-06,
   9.912050700222608e-07,
   6.691512908219011e-07,
   5.152484163772897e-07,
   4.32310315545692e-07,
   2.8608965862986224e-07,
   2.3783076130712288e-07
  ],
  "state_rebuild_rel_err_max": 5.324042831489351e-06
 },
 "final": {
  "grad_rel_err": 8.689006283887884e-05,
  "grad_cosine": 0.9999999962297621,
  "state_rebuild_rel_err_per_layer": [
   6.500593735836446e-05,
 